### PR TITLE
feat: lua functions and closure support for '*expr'/'*func' options

### PR DESCRIPTION
**Pull request by famiu**
Adds support for Lua functions and Vim script closures for `*expr`/`*func` options. (eg: `'indentexpr'`. `'tagfunc'`, etc.)

Very WIP and currently in a broken and uncompilable state.

### CODE REVIEW

File: src/nvim/eval/typval_defs.h [hunk 1]
```diff
@@ -0,0 +1,396 @@
...
+/// List actions log
+struct list_log {
+  ListLog *next;  ///< Next chunk or NULL.
+  size_t capacity;  ///< Number of entries in current chunk.
+  size_t size;  ///< Current chunk size.
+  ListLogEntry entries[];  ///< Actual log entries.
+};
+
+extern ListLog *list_log_first;  ///< First list log chunk, NULL if missing
+extern ListLog *list_log_last;  ///< Last list log chunk
```
Comment (zeertzjq) [inline src/nvim/eval/typval_defs.h:393]: These two should be in `typval.h`

File: src/nvim/eval.c [hunk 2]
```diff
@@ -3676,37 +3677,22 @@ int get_option_tv(const char **const arg, typval_T *const rettv, const bool eval
...
-      rettv->v_type = VAR_STRING;
-      rettv->vval.v_string = stringval;
-    }
-  } else if (working && (opt_type == gov_hidden_bool
-                         || opt_type == gov_hidden_number
-                         || opt_type == gov_hidden_string)) {
+    Error err;
+    object_to_vim(val, rettv, &err);
+    assert(err.type == kErrorTypeNone);
+  } else if (working && hidden) {
```
Comment (lewis6991) [inline src/nvim/eval.c:3695]: This is me being picky, but I think refactors like this deserve their own PR which should be merged separately.

Comment (famiu) [reply src/nvim/eval.c:3695]: This is actually necessary for the PR

Comment (lewis6991) [reply src/nvim/eval.c:3695]: I only mentioned because we specifically talked about this on matrix.

This is a big PR and will be hard to review, so even though it is unfinished, you should appreciate early feedback.
